File: Leetcode/0055-jump-game/solution.go
package main

import "fmt"

// ============================================================
// 0055. Jump Game
// https://leetcode.com/problems/jump-game/
// Difficulty: Medium
// Tags: Array, Dynamic Programming, Greedy
// ============================================================

// canJump — Optimal Solution (Greedy)
// Approach: maintain the farthest reachable index. If current index ever
//   exceeds it we are stuck; if farthest >= n-1 we have won.
// Time:  O(n)
// Space: O(1)
func canJump(nums []int) bool {
	farthest := 0
	for i := 0; i < len(nums); i++ {
		if i > farthest {
			return false
		}
		if i+nums[i] > farthest {
			farthest = i + nums[i]
		}
		if farthest >= len(nums)-1 {
			return true
		}
	}
	return true
}

// canJumpDP — Bottom-Up DP (right-to-left)
// Time:  O(n)
// Space: O(1)
func canJumpDP(nums []int) bool {
	n := len(nums)
	lastGood := n - 1
	for i := n - 2; i >= 0; i-- {
		if i+nums[i] >= lastGood {
			lastGood = i
		}
	}
	return lastGood == 0
}

// canJumpMemo — Top-Down DP with memoization
// Time:  O(n^2)
// Space: O(n)
func canJumpMemo(nums []int) bool {
	n := len(nums)
	dp := make([]int8, n)
	var solve func(i int) bool
	solve = func(i int) bool {
		if i >= n-1 {
			return true
		}
		if dp[i] != 0 {
			return dp[i] == 1
		}
		furthest := i + nums[i]
		if furthest >= n-1 {
			furthest = n - 1
		}
		for j := i + 1; j <= furthest; j++ {
			if solve(j) {
				dp[i] = 1
				return true
			}
		}
		dp[i] = -1
		return false
	}
	return solve(0)
}

// canJumpBrute — Pure recursion (TLE for large n)
// Time:  O(2^n)
// Space: O(n)
func canJumpBrute(nums []int) bool {
	var rec func(i int) bool
	rec = func(i int) bool {
		if i >= len(nums)-1 {
			return true
		}
		furthest := i + nums[i]
		if furthest >= len(nums)-1 {
			furthest = len(nums) - 1
		}
		for j := i + 1; j <= furthest; j++ {
			if rec(j) {
				return true
			}
		}
		return false
	}
	return rec(0)
}

// ============================================================
// Test Cases
// ============================================================

func main() {
	passed, failed := 0, 0
	test := func(name string, got, expected bool) {
		if got == expected {
			fmt.Printf("PASS: %s\n", name)
			passed++
		} else {
			fmt.Printf("FAIL: %s\n  Got:      %v\n  Expected: %v\n", name, got, expected)
			failed++
		}
	}

	type tc struct {
		name     string
		nums     []int
		expected bool
	}
	cases := []tc{
		{"Example 1", []int{2, 3, 1, 1, 4}, true},
		{"Example 2", []int{3, 2, 1, 0, 4}, false},
		{"Single zero", []int{0}, true},
		{"Single positive", []int{5}, true},
		{"Cannot move", []int{0, 1}, false},
		{"Big first jump", []int{100, 0, 0, 0}, true},
		{"All ones", []int{1, 1, 1, 1, 1}, true},
		{"Edge size 2 reachable", []int{1, 0}, true},
		{"Two zeros block", []int{2, 0, 0, 1}, false},
		{"Just barely", []int{2, 0, 0}, true},
		{"Need two jumps", []int{1, 2, 3}, true},
		{"All zeros except first", []int{4, 0, 0, 0, 0}, true},
		{"Long zeros chain", []int{1, 1, 0, 1}, false},
		{"Final element irrelevant value", []int{2, 1, 0, 0, 0}, false},
	}

	fmt.Println("=== Greedy ===")
	for _, c := range cases {
		test(c.name, canJump(c.nums), c.expected)
	}

	fmt.Println("\n=== Bottom-Up DP ===")
	for _, c := range cases {
		test("DP "+c.name, canJumpDP(c.nums), c.expected)
	}

	fmt.Println("\n=== Top-Down DP (Memo) ===")
	for _, c := range cases {
		test("Memo "+c.name, canJumpMemo(c.nums), c.expected)
	}

	// Brute only for small inputs
	fmt.Println("\n=== Brute Force (small only) ===")
	for _, c := range cases {
		if len(c.nums) > 12 {
			continue
		}
		test("Brute "+c.name, canJumpBrute(c.nums), c.expected)
	}

	fmt.Printf("\nResults: %d passed, %d failed, %d total\n", passed, failed, passed+failed)
}

File: Leetcode/0055-jump-game/solution.java
import java.util.*;

/**
 * 0055. Jump Game
 * https://leetcode.com/problems/jump-game/
 * Difficulty: Medium
 * Tags: Array, Dynamic Programming, Greedy
 */
class Solution {

    /**
     * Optimal Solution (Greedy).
     * Time:  O(n)
     * Space: O(1)
     */
    public boolean canJump(int[] nums) {
        int farthest = 0;
        for (int i = 0; i < nums.length; i++) {
            if (i > farthest) return false;
            farthest = Math.max(farthest, i + nums[i]);
            if (farthest >= nums.length - 1) return true;
        }
        return true;
    }

    /**
     * Bottom-Up DP.
     * Time:  O(n)
     * Space: O(1)
     */
    public boolean canJumpDP(int[] nums) {
        int n = nums.length;
        int lastGood = n - 1;
        for (int i = n - 2; i >= 0; i--) {
            if (i + nums[i] >= lastGood) lastGood = i;
        }
        return lastGood == 0;
    }

    /**
     * Top-Down DP with memoization.
     * Time:  O(n^2)
     * Space: O(n)
     */
    public boolean canJumpMemo(int[] nums) {
        byte[] dp = new byte[nums.length];
        return solveMemo(nums, dp, 0);
    }

    private boolean solveMemo(int[] nums, byte[] dp, int i) {
        int n = nums.length;
        if (i >= n - 1) return true;
        if (dp[i] != 0) return dp[i] == 1;
        int furthest = Math.min(i + nums[i], n - 1);
        for (int j = i + 1; j <= furthest; j++) {
            if (solveMemo(nums, dp, j)) {
                dp[i] = 1;
                return true;
            }
        }
        dp[i] = -1;
        return false;
    }

    /**
     * Brute force (TLE for large n).
     * Time:  O(2^n)
     * Space: O(n)
     */
    public boolean canJumpBrute(int[] nums) {
        return rec(nums, 0);
    }
    private boolean rec(int[] nums, int i) {
        if (i >= nums.length - 1) return true;
        int furthest = Math.min(i + nums[i], nums.length - 1);
        for (int j = i + 1; j <= furthest; j++) {
            if (rec(nums, j)) return true;
        }
        return false;
    }

    // ============================================================
    // Test Cases
    // ============================================================

    static int passed = 0, failed = 0;

    static void test(String name, boolean got, boolean expected) {
        if (got == expected) {
            System.out.println("PASS: " + name);
            passed++;
        } else {
            System.out.println("FAIL: " + name);
            System.out.println("  Got:      " + got);
            System.out.println("  Expected: " + expected);
            failed++;
        }
    }

    public static void main(String[] args) {
        Solution sol = new Solution();
        Object[][] cases = {
            {"Example 1", new int[]{2, 3, 1, 1, 4}, true},
            {"Example 2", new int[]{3, 2, 1, 0, 4}, false},
            {"Single zero", new int[]{0}, true},
            {"Single positive", new int[]{5}, true},
            {"Cannot move", new int[]{0, 1}, false},
            {"Big first jump", new int[]{100, 0, 0, 0}, true},
            {"All ones", new int[]{1, 1, 1, 1, 1}, true},
            {"Edge size 2 reachable", new int[]{1, 0}, true},
            {"Two zeros block", new int[]{2, 0, 0, 1}, false},
            {"Just barely", new int[]{2, 0, 0}, true},
            {"Need two jumps", new int[]{1, 2, 3}, true},
            {"All zeros except first", new int[]{4, 0, 0, 0, 0}, true},
            {"Long zeros chain", new int[]{1, 1, 0, 1}, false},
            {"Final element irrelevant", new int[]{2, 1, 0, 0, 0}, false},
        };

        System.out.println("=== Greedy ===");
        for (Object[] c : cases) test((String) c[0], sol.canJump((int[]) c[1]), (boolean) c[2]);

        System.out.println("\n=== Bottom-Up DP ===");
        for (Object[] c : cases) test("DP " + c[0], sol.canJumpDP((int[]) c[1]), (boolean) c[2]);

        System.out.println("\n=== Top-Down DP (Memo) ===");
        for (Object[] c : cases) test("Memo " + c[0], sol.canJumpMemo((int[]) c[1]), (boolean) c[2]);

        System.out.println("\n=== Brute Force (small only) ===");
        for (Object[] c : cases) {
            if (((int[]) c[1]).length > 12) continue;
            test("Brute " + c[0], sol.canJumpBrute((int[]) c[1]), (boolean) c[2]);
        }

        System.out.printf("%nResults: %d passed, %d failed, %d total%n",
            passed, failed, passed + failed);
    }
}
